README.md
# MudParser

[![Python 3.10+](https://img.shields.io/badge/python-3.10+-blue.svg)](https://www.python.org/downloads/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![RFC 8520](https://img.shields.io/badge/RFC-8520-green.svg)](https://datatracker.ietf.org/doc/html/rfc8520)
[![RFC 9761](https://img.shields.io/badge/RFC-9761-green.svg)](https://datatracker.ietf.org/doc/html/rfc9761)
[![Tests](https://img.shields.io/badge/tests-90%20passed-brightgreen.svg)]()
[![Code style: ruff](https://img.shields.io/badge/code%20style-ruff-000000.svg)](https://github.com/astral-sh/ruff)

A production-grade Python library for parsing, validating, and exporting **Manufacturer Usage Description (MUD)** profiles as defined in [RFC 8520](https://datatracker.ietf.org/doc/html/rfc8520) with full support for [RFC 9761](https://datatracker.ietf.org/doc/html/rfc9761) TLS/DTLS profiles.

MUD profiles allow IoT device manufacturers to formally describe the network behavior their devices require, enabling network administrators to automatically configure access control policies and significantly reduce the attack surface of IoT deployments.

---

## Table of Contents

- [What is MUD?](#what-is-mud)
- [Features](#features)
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Detailed Usage Guide](#detailed-usage-guide)
  - [Parsing MUD Files](#parsing-mud-files)
  - [Accessing Profile Data](#accessing-profile-data)
  - [Working with ACLs and ACEs](#working-with-acls-and-aces)
  - [Validation](#validation)
  - [Exporting to Firewall Rules](#exporting-to-firewall-rules)
- [Command Line Interface](#command-line-interface)
- [Interactive Web Demo](#interactive-web-demo)
- [RFC Compliance](#rfc-compliance)
- [Sample MUD Profiles](#sample-mud-profiles)
- [API Reference](#api-reference)
- [Architecture](#architecture)
- [Development](#development)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)
- [License](#license)
- [Acknowledgements](#acknowledgements)

---

## What is MUD?

**Manufacturer Usage Description (MUD)** is an IETF standard ([RFC 8520](https://datatracker.ietf.org/doc/html/rfc8520)) that provides a formal way for IoT device manufacturers to describe the intended network behavior of their devices.

### The Problem MUD Solves

IoT devices often have well-defined, limited network communication patterns. However, without formal descriptions:
- Network administrators don't know what traffic is legitimate
- Devices may be over-permissioned on the network
- Compromised devices can be used to attack other systems
- Manual firewall configuration is error-prone and time-consuming

### How MUD Works

```
┌─────────────┐     ┌─────────────┐     ┌─────────────┐
│  IoT Device │────>│ MUD Manager │────>│  Firewall   │
│  (sends URL)│     │  (fetches)  │     │  (enforces) │
└─────────────┘     └──────┬──────┘     └─────────────┘
                          │
                          v
                   ┌─────────────┐
                   │  MUD File   │
                   │  (on web)   │
                   └─────────────┘
```

1. **Device Announcement**: IoT device announces its MUD URL (via DHCP, LLDP, or 802.1X)
2. **Profile Fetch**: MUD manager fetches the MUD file from the URL
3. **Policy Generation**: MUD file is parsed and converted to firewall rules
4. **Enforcement**: Rules are applied to restrict device communication

### MUD File Structure

A MUD file is a JSON document containing:
- **Metadata**: Device info, update timestamps, support status
- **Access Control Lists (ACLs)**: Network communication rules
- **Policy References**: Which ACLs apply to inbound/outbound traffic

---

## Features

### Core Capabilities

| Feature | Description |
|---------|-------------|
| **RFC 8520 Compliant** | Full implementation of the MUD specification |
| **RFC 9761 Support** | TLS/DTLS profile extensions for secure communications |
| **Modern Python** | Built for Python 3.10+ with type hints and Pydantic v2 |
| **Multiple Input Sources** | Parse from files, strings, URLs, or dictionaries |
| **Async Support** | Asynchronous URL fetching for high-performance applications |

### Validation

| Feature | Description |
|---------|-------------|
| **Schema Validation** | Validates against RFC 8520 JSON schema |
| **Semantic Validation** | Cross-reference checking, constraint validation |
| **Severity Levels** | Errors, warnings, and informational messages |
| **Detailed Reporting** | Path-specific error messages with context |

### Export Formats

| Format | Platform | Use Case |
|--------|----------|----------|
| **iptables** | Linux | Traditional Linux firewall |
| **nftables** | Linux | Modern Linux firewall (netfilter) |
| **Cisco ACL** | Cisco IOS | Enterprise routers and switches |
| **pfSense** | BSD/pfSense | BSD-based firewalls |
| **JSON** | Any | Re-serialization, APIs |
| **YAML** | Any | Human-readable configuration |

### User Interfaces

| Interface | Description |
|-----------|-------------|
| **Python API** | Full-featured library for programmatic access |
| **CLI** | Rich command-line interface with syntax highlighting |
| **Web Demo** | Interactive Streamlit application |

---

## Installation

### Requirements

- **Python 3.10 or higher** (uses modern features like match/case, union types)
- **pip** package manager

### Basic Installation

```bash
pip install mudparser
```

### Installation with Extras

```bash
# Development tools (pytest, mypy, ruff)
pip install mudparser[dev]

# Documentation tools (mkdocs, mkdocstrings)
pip install mudparser[docs]

# Demo application (streamlit)
pip install mudparser[demo]

# All extras
pip install mudparser[all]
```

### Installation from Source

```bash
# Clone the repository
git clone https://github.com/elmiomar/mudparser.git
cd mudparser

# Install in development mode
pip install -e .

# Or with all extras
pip install -e ".[all]"
```

### Verify Installation

```bash
# Check version
mudparser --version

# Run a quick test
mudparser --help
```

---

## Quick Start

### 30-Second Example

```python
from mudparser import MUDParser

# Parse a MUD file
parser = MUDParser.from_file("device.mud.json")

# Get device info
print(f"Device: {parser.mud.systeminfo}")
print(f"Rules: {parser.get_summary()['total_rules']}")

# Generate firewall rules
rules = parser.export.to_iptables(device_ip="192.168.1.100")
print(rules)
```

### CLI Quick Start

```bash
# View profile information
mudparser info device.mud.json

# Validate a profile
mudparser validate device.mud.json

# Export to iptables
mudparser export device.mud.json -f iptables -d 192.168.1.100
```

---

## Detailed Usage Guide

### Parsing MUD Files

#### From a Local File

```python
from mudparser import MUDParser

# Parse from file path (string or Path object)
parser = MUDParser.from_file("path/to/device.mud.json")

# With pathlib
from pathlib import Path
parser = MUDParser.from_file(Path("data") / "device.mud.json")
```

#### From a JSON String

```python
from mudparser import MUDParser

json_content = '''
{
    "ietf-mud:mud": {
        "mud-version": 1,
        "mud-url": "https://example.com/device.mud.json",
        "last-update": "2024-01-15T00:00:00Z",
        "cache-validity": 48,
        "is-supported": true,
        "systeminfo": "My IoT Device",
        "from-device-policy": { ... },
        "to-device-policy": { ... }
    },
    "ietf-access-control-list:access-lists": { ... }
}
'''

parser = MUDParser.from_string(json_content, source="inline")
```

#### From a URL

```python
from mudparser import MUDParser

# Synchronous fetch
parser = MUDParser.from_url("https://iotanalytics.unsw.edu.au/mud/amazonEchoMud.json")

# Asynchronous fetch (for high-performance applications)
import asyncio

async def fetch_profile():
    parser = await MUDParser.from_url_async(
        "https://iotanalytics.unsw.edu.au/mud/amazonEchoMud.json"
    )
    return parser

parser = asyncio.run(fetch_profile())
```

#### From a Dictionary

```python
from mudparser import MUDParser

data = {
    "ietf-mud:mud": {
        "mud-version": 1,
        "mud-url": "https://example.com/device.mud.json",
        "last-update": "2024-01-15T00:00:00Z",
        "cache-validity": 48,
        "is-supported": True,
        "systeminfo": "My IoT Device",
        "from-device-policy": {
            "access-lists": {
                "access-list": [{"name": "from-device-acl"}]
            }
        },
        "to-device-policy": {
            "access-lists": {
                "access-list": [{"name": "to-device-acl"}]
            }
        }
    },
    "ietf-access-control-list:access-lists": {
        "acl": [
            # ... ACL definitions
        ]
    }
}

parser = MUDParser.from_dict(data, source="programmatic")
```

### Accessing Profile Data

#### MUD Container (Metadata)

```python
from mudparser import MUDParser

parser = MUDParser.from_file("device.mud.json")

# Core metadata (always present)
print(f"MUD Version: {parser.mud.mud_version}")          # 1
print(f"MUD URL: {parser.mud.mud_url}")                  # https://...
print(f"Last Update: {parser.mud.last_update}")          # datetime object
print(f"Cache Validity: {parser.mud.cache_validity}h")   # Hours (1-168)
print(f"Is Supported: {parser.mud.is_supported}")        # True/False
print(f"System Info: {parser.mud.systeminfo}")           # Device description

# Optional metadata
print(f"Manufacturer: {parser.mud.mfg_name}")            # May be None
print(f"Model: {parser.mud.model_name}")                 # May be None
print(f"Firmware: {parser.mud.firmware_rev}")            # May be None
print(f"Software: {parser.mud.software_rev}")            # May be None
print(f"Documentation: {parser.mud.documentation}")      # URL or None
print(f"MUD Signature: {parser.mud.mud_signature}")      # Signature URL
print(f"Extensions: {parser.mud.extensions}")            # List of extensions
```

#### Profile Summary

```python
summary = parser.get_summary()

print(f"URL: {summary['url']}")
print(f"Version: {summary['version']}")
print(f"Device: {summary['systeminfo']}")
print(f"Last Update: {summary['last_update']}")
print(f"Cache Validity: {summary['cache_validity_hours']} hours")
print(f"Supported: {summary['is_supported']}")
print(f"Manufacturer: {summary['manufacturer']}")
print(f"Model: {summary['model']}")

# Rule counts
print(f"Total ACLs: {summary['total_acls']}")
print(f"From-Device ACLs: {summary['from_device_acls']}")
print(f"To-Device ACLs: {summary['to_device_acls']}")
print(f"Total Rules: {summary['total_rules']}")
print(f"From-Device Rules: {summary['from_device_rules']}")
print(f"To-Device Rules: {summary['to_device_rules']}")

# Network resources
print(f"DNS Names: {summary['dns_names']}")
print(f"TCP Ports: {summary['ports']['tcp']}")
print(f"UDP Ports: {summary['ports']['udp']}")
```

#### Extracting Network Resources

```python
# Get all DNS names referenced in the profile
dns_names = parser.get_dns_names()
print("DNS Names:")
for name in sorted(dns_names):
    print(f"  - {name}")

# Get all ports used
ports = parser.get_ports()
print(f"TCP Ports: {sorted(ports['tcp'])}")
print(f"UDP Ports: {sorted(ports['udp'])}")

# Get all ACE entries with direction
for direction, entry in parser.get_all_entries():
    print(f"[{direction}] {entry.name}: {entry.get_description(direction)}")
```

### Working with ACLs and ACEs

#### Iterating Over ACLs

```python
# Get all ACLs
for acl in parser.profile.acls.acl:
    print(f"\nACL: {acl.name}")
    print(f"  Type: {acl.acl_type.value}")  # ipv4-acl-type, ipv6-acl-type, etc.
    print(f"  Rules: {len(acl)}")

# Get ACLs by direction
print("\n=== FROM-DEVICE (Outbound) ===")
for acl in parser.get_from_device_acls():
    print(f"  {acl.name}: {len(acl)} rules")

print("\n=== TO-DEVICE (Inbound) ===")
for acl in parser.get_to_device_acls():
    print(f"  {acl.name}: {len(acl)} rules")
```

#### Working with ACL Entries (ACEs)

```python
for acl in parser.profile.acls.acl:
    print(f"\nACL: {acl.name}")

    for entry in acl.entries:
        # Basic info
        print(f"  Rule: {entry.name}")

        # Action
        if entry.is_accept():
            print("    Action: ALLOW")
        else:
            print("    Action: DENY")

        # Get human-readable description
        desc = entry.get_description(direction="from")
        print(f"    Description: {desc}")

        # Access match conditions
        matches = entry.matches

        # Protocol information
        protocol = matches.get_protocol()
        print(f"    Protocol: {protocol}")

        # DNS names in this rule
        dns_names = matches.get_dns_names()
        if dns_names:
            print(f"    DNS Names: {dns_names}")
```

#### Accessing Match Details

```python
for acl in parser.profile.acls.acl:
    for entry in acl.entries:
        matches = entry.matches

        # IPv4 matches
        if matches.ipv4:
            print(f"  IPv4 Protocol: {matches.ipv4.protocol}")
            if matches.ipv4.src_network:
                print(f"  Source Network: {matches.ipv4.src_network}")
            if matches.ipv4.dst_network:
                print(f"  Dest Network: {matches.ipv4.dst_network}")
            if matches.ipv4.src_dnsname:
                print(f"  Source DNS: {matches.ipv4.src_dnsname}")
            if matches.ipv4.dst_dnsname:
                print(f"  Dest DNS: {matches.ipv4.dst_dnsname}")

        # IPv6 matches
        if matches.ipv6:
            print(f"  IPv6 Protocol: {matches.ipv6.protocol}")
            if matches.ipv6.dst_network:
                print(f"  Dest Network: {matches.ipv6.dst_network}")

        # TCP matches
        if matches.tcp:
            if matches.tcp.src_port:
                print(f"  TCP Source Port: {matches.tcp.src_port}")
            if matches.tcp.dst_port:
                print(f"  TCP Dest Port: {matches.tcp.dst_port}")
            if matches.tcp.direction_initiated:
                print(f"  Direction: {matches.tcp.direction_initiated.value}")

        # UDP matches
        if matches.udp:
            if matches.udp.src_port:
                print(f"  UDP Source Port: {matches.udp.src_port}")
            if matches.udp.dst_port:
                print(f"  UDP Dest Port: {matches.udp.dst_port}")

        # ICMP matches
        if matches.icmp:
            print(f"  ICMP Type: {matches.icmp.type}, Code: {matches.icmp.code}")

        # Ethernet matches
        if matches.eth:
            if matches.eth.src_mac:
                print(f"  Source MAC: {matches.eth.src_mac}")
            if matches.eth.dst_mac:

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2026-02-04

### Added

#### Core Features
- **RFC 8520 Full Compliance** - Complete implementation of Manufacturer Usage Description specification
- **RFC 9761 Support** - TLS/DTLS profile extensions for secure communications
- **Modern Python 3.10+** - Leveraging match/case statements, union types, and modern syntax
- **Pydantic v2 Models** - Type-safe data models with automatic validation

#### Parsing
- `MUDParser.from_file()` - Parse MUD profiles from local files
- `MUDParser.from_string()` - Parse from JSON strings
- `MUDParser.from_url()` - Fetch and parse from HTTP/HTTPS URLs
- `MUDParser.from_url_async()` - Async URL fetching support
- `MUDParser.from_dict()` - Parse from Python dictionaries

#### Data Models
- `MUDProfile` - Complete MUD container model with all RFC 8520 fields
- `AccessControlList` - ACL model per RFC 8519
- `AccessControlEntry` - ACE model with full match support
- **Match Types:**
  - `IPv4Match` - IPv4 header matching (protocol, addresses, DSCP, TTL)
  - `IPv6Match` - IPv6 header matching (protocol, addresses, flow label)
  - `TCPMatch` - TCP matching (ports, flags, direction-initiated)
  - `UDPMatch` - UDP matching (ports)
  - `ICMPMatch` - ICMP/ICMPv6 matching (type, code)
  - `EthernetMatch` - Ethernet frame matching (MAC addresses, ethertype)
  - `MUDMatch` - MUD-specific matching (manufacturer, controller, local-networks)
- **Port Operators:** eq, lt, gt, neq, range
- **TLS Models (RFC 9761):**
  - `TLSProfile` - TLS profile constraints
  - `DTLSProfile` - DTLS profile constraints
  - `CipherSuiteConstraints` - Allowed/forbidden cipher suites
  - `SPKIPinSet` - Certificate pinning support

#### Validation
- `MUDValidator` - Comprehensive validation engine
- `ValidationResult` - Detailed validation results with severity levels
- `ValidationIssue` - Individual issues with path and context
- Schema validation against RFC 8520
- Semantic validation (cross-references, constraints)
- Configurable severity levels (ERROR, WARNING, INFO)

#### Export Capabilities
- **iptables** - Linux firewall rules with chain management
- **nftables** - Modern Linux firewall configuration
- **Cisco ACL** - Cisco IOS access control lists
- **pfSense** - BSD firewall XML rules
- **JSON** - Re-serialized MUD profile
- **YAML** - Human-readable format

#### Command Line Interface
- `mudparser validate` - Validate MUD profiles with detailed reporting
- `mudparser info` - Display profile summary and metadata
- `mudparser rules` - Show access control rules in human-readable format
- `mudparser export` - Export to firewall formats
- `mudparser fetch` - Fetch profiles from URLs
- `mudparser diff` - Compare two MUD profiles
- `mudparser demo` - Launch interactive Streamlit demo
- Rich terminal output with syntax highlighting
- JSON/YAML output options

#### Interactive Demo
- Streamlit web application
- File upload and JSON paste support
- Visual profile overview
- Interactive rule browser
- Validation report display
- One-click export to all formats
- Download generated rules

#### Documentation
- MkDocs with Material theme
- Comprehensive user guide
- API reference documentation
- RFC compliance documentation
- Example code for all features
- Installation and quickstart guides

#### Testing
- pytest test suite
- Comprehensive test fixtures
- Model validation tests
- Parser tests
- Exporter tests
- CLI tests

### Changed

#### Breaking Changes
- **Package Structure** - Moved to `src/` layout (`src/mudparser/`)
- **Python Version** - Now requires Python 3.10+ (previously 3.5+)
- **Import Paths** - Models now under `mudparser.models.*`
- **API Changes:**
  - `MUDParser` is now the main entry point (replaces old `mudparser` module)
  - Validation returns `ValidationResult` instead of simple boolean
  - Export methods require explicit device IP for firewall formats

#### Improvements
- Complete rewrite with modern Python idioms
- Type hints throughout the codebase
- Pydantic v2 for data validation (replaces manual parsing)
- Async support for URL fetching
- Better error messages with context

### Removed
- `setup.py` - Replaced with `pyproject.toml`
- Old `mudparser/` package structure
- Python 2.x and Python 3.5-3.9 support

### Fixed
- Proper handling of optional MUD fields
- Correct parsing of port range operators
- DNS name extraction from all match types
- ACL cross-reference validation

## [1.0.0-alpha] - Previous

### Added
- Initial alpha implementation
- Basic MUD file parsing
- Simple profile representation

### Notes
- Alpha version, not production-ready
- Limited RFC 8520 compliance
- No validation or export capabilities

---

## Migration Guide (1.x to 2.0)

### Installation

```bash
# Old
pip install mudparser
# or
python setup.py install

# New
pip install mudparser
# or for development
pip install -e ".[dev]"
```

### Basic Usage

```python
# Old (1.x)
from mudparser import MUDParser
parser = MUDParser()
parser.parse_file("device.mud.json")

# New (2.0)
from mudparser import MUDParser
parser = MUDParser.from_file("device.mud.json")
```

### Accessing Data

```python
# Old (1.x)
mud_url = parser.mud_url
system_info = parser.systeminfo

# New (2.0)
mud_url = parser.mud.mud_url
system_info = parser.mud.systeminfo
```

### Validation

```python
# Old (1.x)
# No validation available

# New (2.0)
errors = parser.validate()
# or detailed
from mudparser.validator import MUDValidator
validator = MUDValidator()
result = validator.validate(parser.profile)
```

### Export

```python
# Old (1.x)
# No export available

# New (2.0)
rules = parser.export.to_iptables(device_ip="192.168.1.100")
```

---

[2.0.0]: https://github.com/elmiomar/mudparser/releases/tag/v2.0.0

demo/streamlit_app.py
"""
MudParser Interactive Demo Application.

A Streamlit-based web application for exploring MUD profiles.

Run with:
    streamlit run demo/streamlit_app.py

Or via CLI:
    mudparser demo
"""

import json
import sys
from pathlib import Path

import streamlit as st

# Add src to path for development
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from mudparser import MUDParser, __version__
from mudparser.exceptions import MUDParserError
from mudparser.exporters import ExportFormat
from mudparser.validator import MUDValidator, ValidationSeverity

# Page configuration
st.set_page_config(
    page_title="MudParser Demo",
    page_icon="🔒",
    layout="wide",
    initial_sidebar_state="expanded",
)

# Custom CSS
st.markdown("""
<style>
    .stTabs [data-baseweb="tab-list"] {
        gap: 24px;
    }
    .stTabs [data-baseweb="tab"] {
        padding: 10px 20px;
    }
    .validation-error {
        color: #ff4b4b;
        font-weight: bold;
    }
    .validation-warning {
        color: #ffa500;
    }
    .validation-info {
        color: #00a2ff;
    }
</style>
""", unsafe_allow_html=True)


def main():
    """Main application entry point."""
    st.title("🔒 MudParser Demo")
    st.markdown(f"*RFC 8520 & RFC 9761 Compliant MUD Profile Parser* — v{__version__}")

    # Sidebar
    with st.sidebar:
        st.header("📁 Load MUD Profile")

        load_method = st.radio(
            "Choose input method:",
            ["Upload File", "Paste JSON", "Sample Profile"],
            label_visibility="collapsed",
        )

        parser = None

        if load_method == "Upload File":
            uploaded_file = st.file_uploader(
                "Upload MUD JSON file",
                type=["json"],
                help="Upload a MUD profile in JSON format",
            )
            if uploaded_file:
                try:
                    content = uploaded_file.read().decode("utf-8")
                    parser = MUDParser.from_string(content, source=uploaded_file.name)
                    st.success(f"Loaded: {uploaded_file.name}")
                except MUDParserError as e:
                    st.error(f"Error: {e}")

        elif load_method == "Paste JSON":
            json_input = st.text_area(
                "Paste MUD JSON:",
                height=200,
                placeholder='{"ietf-mud:mud": {...}}',
            )
            if json_input:
                try:
                    parser = MUDParser.from_string(json_input, source="pasted")
                    st.success("JSON parsed successfully!")
                except MUDParserError as e:
                    st.error(f"Error: {e}")

        else:  # Sample Profile
            sample = get_sample_profile()
            parser = MUDParser.from_dict(sample, source="sample")
            st.info("Using sample Amazon Echo profile")

        st.divider()
        st.markdown("### About")
        st.markdown("""
        MudParser parses MUD (Manufacturer Usage Description) profiles
        that define network access policies for IoT devices.

        [Documentation](https://elmiomar.github.io/mudparser) |
        [GitHub](https://github.com/elmiomar/mudparser)
        """)

    # Main content
    if parser:
        display_profile(parser)
    else:
        st.info("👈 Load a MUD profile from the sidebar to get started.")
        display_instructions()


def display_profile(parser: MUDParser):
    """Display the parsed MUD profile."""
    # Tabs
    tab1, tab2, tab3, tab4, tab5 = st.tabs([
        "📋 Overview",
        "🔍 Rules",
        "✅ Validation",
        "📤 Export",
        "📄 Raw JSON",
    ])

    with tab1:
        display_overview(parser)

    with tab2:
        display_rules(parser)

    with tab3:
        display_validation(parser)

    with tab4:
        display_export(parser)

    with tab5:
        display_json(parser)


def display_overview(parser: MUDParser):
    """Display profile overview."""
    st.header("Profile Overview")

    summary = parser.get_summary()

    # Device info
    col1, col2 = st.columns(2)

    with col1:
        st.subheader("Device Information")
        st.metric("Device Name", summary.get("systeminfo") or "Unknown")
        st.metric("Supported", "Yes" if summary["is_supported"] else "No")
        if summary.get("manufacturer"):
            st.metric("Manufacturer", summary["manufacturer"])

    with col2:
        st.subheader("MUD Metadata")
        st.metric("MUD Version", summary["version"])
        st.metric("Cache Validity", f"{summary['cache_validity_hours']} hours")
        st.text(f"URL: {summary['url']}")

    # Rules summary
    st.subheader("Access Control Summary")
    col1, col2, col3 = st.columns(3)

    with col1:
        st.metric("Total Rules", summary["total_rules"])
    with col2:
        st.metric("From-Device (Outbound)", summary["from_device_rules"])
    with col3:
        st.metric("To-Device (Inbound)", summary["to_device_rules"])

    # Resources
    col1, col2 = st.columns(2)

    with col1:
        st.subheader("Referenced DNS Names")
        if summary["dns_names"]:
            for name in sorted(summary["dns_names"]):
                st.code(name)
        else:
            st.info("No DNS names referenced")

    with col2:
        st.subheader("Referenced Ports")
        if summary["ports"]["tcp"]:
            st.write(f"**TCP:** {', '.join(map(str, sorted(summary['ports']['tcp'])))}")
        if summary["ports"]["udp"]:
            st.write(f"**UDP:** {', '.join(map(str, sorted(summary['ports']['udp'])))}")
        if not summary["ports"]["tcp"] and not summary["ports"]["udp"]:
            st.info("No specific ports referenced")


def display_rules(parser: MUDParser):
    """Display access control rules."""
    st.header("Access Control Rules")

    # From-device rules
    st.subheader("📤 From-Device Policy (Outbound)")
    for acl in parser.get_from_device_acls():
        with st.expander(f"ACL: {acl.name} ({acl.acl_type.value})", expanded=True):
            for entry in acl.entries:
                display_ace(entry, "from")

    # To-device rules
    st.subheader("📥 To-Device Policy (Inbound)")
    for acl in parser.get_to_device_acls():
        with st.expander(f"ACL: {acl.name} ({acl.acl_type.value})", expanded=True):
            for entry in acl.entries:
                display_ace(entry, "to")


def display_ace(entry, direction: str):
    """Display a single ACE entry."""
    action_color = "green" if entry.is_accept() else "red"
    action_text = "ALLOW" if entry.is_accept() else "DENY"

    desc = entry.get_description(direction)

    st.markdown(f"""
    <div style="padding: 8px; background-color: #f0f2f6; border-radius: 4px; margin: 4px 0;">
        <span style="color: {action_color}; font-weight: bold;">[{action_text}]</span>
        <code>{entry.name}</code>
        <br/>
        <small>{desc}</small>
    </div>
    """, unsafe_allow_html=True)


def display_validation(parser: MUDParser):
    """Display validation results."""
    st.header("Validation Results")

    validator = MUDValidator()
    result = validator.validate(parser.profile)

    # Summary
    if result.is_valid:
        st.success("✅ Profile is valid!")
    else:
        st.error("❌ Profile validation failed")

    col1, col2, col3 = st.columns(3)
    with col1:
        st.metric("Errors", result.error_count)
    with col2:
        st.metric("Warnings", result.warning_count)
    with col3:
        st.metric("Info", len(result.issues) - result.error_count - result.warning_count)

    # Issues list
    if result.issues:
        st.subheader("Issues Found")

        for issue in result.issues:
            if issue.severity == ValidationSeverity.ERROR:
                st.error(f"**ERROR:** {issue.message}")
            elif issue.severity == ValidationSeverity.WARNING:
                st.warning(f"**WARNING:** {issue.message}")
            else:
                st.info(f"**INFO:** {issue.message}")

            if issue.path:
                st.caption(f"Path: `{issue.path}`")


def display_export(parser: MUDParser):
    """Display export options."""
    st.header("Export Profile")

    col1, col2 = st.columns([1, 2])

    with col1:
        export_format = st.selectbox(
            "Export Format",
            [f.value for f in ExportFormat],
            format_func=lambda x: {
                "json": "JSON",
                "yaml": "YAML",
                "iptables": "iptables (Linux)",
                "nftables": "nftables (Linux)",
                "cisco": "Cisco IOS ACL",
                "pfsense": "pfSense XML",
            }.get(x, x),
        )

        # Device IP for firewall formats
        device_ip = None
        if export_format in ["iptables", "nftables", "pfsense"]:
            device_ip = st.text_input(
                "Device IP Address",
                value="192.168.1.100",
                help="Required for firewall rule generation",
            )

        export_button = st.button("Generate Export", type="primary")

    with col2:
        if export_button:
            try:
                kwargs = {}
                if device_ip:
                    kwargs["device_ip"] = device_ip

                output = parser.export.export(export_format, **kwargs)

                # Syntax highlighting
                lang_map = {
                    "json": "json",
                    "yaml": "yaml",
                    "iptables": "bash",
                    "nftables": "text",
                    "cisco": "text",
                    "pfsense": "xml",
                }

                st.code(output, language=lang_map.get(export_format, "text"))

                # Download button
                st.download_button(
                    label="📥 Download",
                    data=output,
                    file_name=f"mud_export.{export_format}",
                    mime="text/plain",
                )

            except Exception as e:
                st.error(f"Export error: {e}")


def display_json(parser: MUDParser):
    """Display raw JSON."""
    st.header("Raw JSON")

    json_str = parser.to_json(indent=2)
    st.code(json_str, language="json")

    st.download_button(
        label="📥 Download JSON",
        data=json_str,
        file_name="mud_profile.json",
        mime="application/json",
    )


def display_instructions():
    """Display getting started instructions."""
    st.header("Getting Started")

    st.markdown("""
    ### What is MUD?

    **MUD (Manufacturer Usage Description)** is an IETF standard (RFC 8520) that allows
    IoT device manufacturers to formally describe the network behavior their devices require.

    ### How to Use This Demo

    1. **Upload** a MUD profile JSON file, or
    2. **Paste** JSON content directly, or
    3. Use the built-in **sample profile**

    Then explore:
    - **Overview** - Device info and summary
    - **Rules** - Access control entries
    - **Validation** - RFC compliance check
    - **Export** - Generate firewall rules
    - **Raw JSON** - View the full profile

    ### Sample MUD Profile Structure

    ```json
    {
      "ietf-mud:mud": {
        "mud-version": 1,
        "mud-url": "https://example.com/device.json",
        "last-update": "2024-01-01T00:00:00Z",
        "cache-validity": 48,
        "is-supported": true,
        "systeminfo": "My IoT Device",
        "from-device-policy": { ... },
        "to-device-policy": { ... }
      },
      "ietf-access-control-list:access-lists": {
        "acl": [ ... ]
      }
    }
    ```
    """)


def get_sample_profile() -> dict:
    """Return a sample MUD profile for demo purposes."""
    return {
        "ietf-mud:mud": {
            "mud-version": 1,
            "mud-url": "https://amazonecho.com/amazonecho",
            "last-update": "2024-01-15T00:00:00Z",
            "cache-validity": 48,
            "is-supported": True,
            "systeminfo": "Amazon Echo (Demo)",
            "from-device-policy": {
                "access-lists": {
                    "access-list": [
                        {"name": "from-ipv4-amazonecho"}
                    ]
                }
            },
            "to-device-policy": {
                "access-lists": {
                    "access-list": [
                        {"name": "to-ipv4-amazonecho"}
                    ]
                }
            },
        },
        "ietf-access-control-list:access-lists": {
            "acl": [
                {
                    "name": "from-ipv4-amazonecho",
                    "type": "ipv4-acl-type",
                    "aces": {
                        "ace": [
                            {
                                "name": "allow-https-amazon",
                                "matches": {
                                    "ipv4": {
                                        "protocol": 6,
                                        "ietf-acldns:dst-dnsname": "dcape-na.amazon.com"
                                    },
                                    "tcp": {
                                        "destination-port": {
                                            "operator": "eq",
                                            "port": 443
                                        }
                                    }
                                },
                                "actions": {"forwarding": "accept"}
                            },
                            {
                                "name": "allow-http-updates",
                                "matches": {
                                    "ipv4": {
                                        "protocol": 6,
                                        "ietf-acldns:dst-dnsname": "softwareupdates.amazon.com"
                                    },
                                    "tcp": {
                                        "destination-port": {
                                            "operator": "eq",
                                            "port": 443
                                        }
                                    }
                                },
                                "actions": {"forwarding": "accept"}
                            }
                        ]
                    }
                },
                {
                    "name": "to-ipv4-amazonecho",
                    "type": "ipv4-acl-type",
                    "aces": {
                        "ace": [
                            {
                                "name": "allow-https-response",
                                "matches": {
                                    "ipv4": {
                                        "protocol": 6,
                                        "ietf-acldns:src-dnsname": "dcape-na.amazon.com"
                                    },
                                    "tcp": {
                                        "source-port": {
                                            "operator": "eq",
                                            "port": 443
                                        }
                                    }
                                },
                                "actions": {"forwarding": "accept"}
                            }
                        ]
                    }
                }
            ]
        }
    }


if __name__ == "__main__":
    main()
